src/types/core.ts

```diff
@@ -43,40 +43,61 @@ export interface VerifierOptions {
  * - A reason to provide further information about the error/invalid/skipped state
  * - Data to provide further information
  */
-export interface VerificationFragment<T = any> {
+export interface VerificationFragment {
   name: string;
   type: VerificationFragmentType;
-  data?: T;
   status: VerificationFragmentStatus;
-  reason?: Reason;
+}
+
+export interface ValidVerificationFragment<Data> extends VerificationFragment {
+  status: "VALID";
+  data: Data;
+}
+export interface InvalidVerificationFragment<Data> extends VerificationFragment {
+  status: "INVALID";
+  reason: Reason;
+  data: Data;
+}
+export interface ErrorVerificationFragment<Data> extends VerificationFragment {
+  status: "ERROR";
+  reason: Reason;
+  data: Data;
+}
+export interface SkippedVerificationFragment extends VerificationFragment {
+  status: "SKIPPED";
+  reason: Reason;
+  // this is useless, it's just to make it easier to work with fragment, otherwise data is not available and need to narrow the type down first
+  data?: undefined;
 }
 export type VerificationFragmentType = "DOCUMENT_INTEGRITY" | "DOCUMENT_STATUS" | "ISSUER_IDENTITY";
 export type VerificationFragmentStatus = "ERROR" | "VALID" | "INVALID" | "SKIPPED";
+export type AllVerificationFragment<Data> =
```

Comment:
As above, to use any since it's unlike any `Data` will be useful here.

Reply:
done